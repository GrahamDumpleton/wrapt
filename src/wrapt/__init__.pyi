import sys

if sys.version_info >= (3, 10):
    from inspect import FullArgSpec
    from types import ModuleType, TracebackType
    from typing import (
        Any,
        Callable,
        Concatenate,
        Generic,
        ParamSpec,
        Protocol,
        TypeVar,
        overload,
    )

    P = ParamSpec("P")
    R = TypeVar("R", covariant=True)

    T = TypeVar("T", bound=Any)

    class Boolean(Protocol):
        def __bool__(self) -> bool: ...

    # ObjectProxy

    class ObjectProxy(Generic[T]):
        __wrapped__: T
        def __init__(self, wrapped: T) -> None: ...

    # CallableObjectProxy

    class CallableObjectProxy(ObjectProxy[T]):
        def __call__(self, *args: Any, **kwargs: Any) -> Any: ...

    # PartialCallableObjectProxy

    class PartialCallableObjectProxy:
        def __init__(
            self, func: Callable[..., Any], *args: Any, **kwargs: Any
        ) -> None: ...
        def __call__(self, *args: Any, **kwargs: Any) -> Any: ...

    def partial(
        func: Callable[..., Any], /, *args: Any, **kwargs: Any
    ) -> Callable[..., Any]: ...

    # WeakFunctionProxy

    class WeakFunctionProxy:
        def __init__(
            self,
            wrapped: Callable[..., Any],
            callback: Callable[..., Any] | None = None,
        ) -> None: ...
        def __call__(self, *args: Any, **kwargs: Any) -> Any: ...

    # FunctionWrapper

    WrappedFunction = Callable[P, R]

    GenericCallableWrapperFunction = Callable[
        [WrappedFunction[P, R], Any, tuple[Any, ...], dict[str, Any]], R
    ]

    ClassMethodWrapperFunction = Callable[
        [type[Any], WrappedFunction[P, R], Any, tuple[Any, ...], dict[str, Any]], R
    ]

    InstanceMethodWrapperFunction = Callable[
        [Any, WrappedFunction[P, R], Any, tuple[Any, ...], dict[str, Any]], R
    ]

    WrapperFunction = (
        GenericCallableWrapperFunction[P, R]
        | ClassMethodWrapperFunction[P, R]
        | InstanceMethodWrapperFunction[P, R]
    )

    class _FunctionWrapperBase(ObjectProxy[WrappedFunction[P, R]]):
        _self_instance: Any
        _self_wrapper: WrapperFunction[P, R]
        _self_enabled: bool | Boolean | Callable[[], bool] | None
        _self_binding: str
        _self_parent: Any
        _self_owner: Any

    class BoundFunctionWrapper(_FunctionWrapperBase[P, R]):
        def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R: ...
        def __get__(
            self, instance: Any, owner: type[Any] | None = None
        ) -> "BoundFunctionWrapper[P, R]": ...

    class FunctionWrapper(_FunctionWrapperBase[P, R]):
        def __init__(
            self,
            wrapped: WrappedFunction[P, R],
            wrapper: WrapperFunction[P, R],
            enabled: bool | Boolean | Callable[[], bool] | None = None,
        ) -> None: ...
        def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R: ...
        def __get__(
            self, instance: Any, owner: type[Any] | None = None
        ) -> BoundFunctionWrapper[P, R]: ...

    # AdapterFactory/adapter_factory()

    class AdapterFactory(Protocol):
        def __call__(
            self, wrapped: Callable[..., Any]
        ) -> str | FullArgSpec | Callable[..., Any]: ...

    def adapter_factory(wrapped: Callable[..., Any]) -> AdapterFactory: ...

    # decorator()

    class Descriptor(Protocol):
        def __get__(self, instance: Any, owner: type[Any] | None = None) -> Any: ...

    class FunctionDecorator(Generic[P, R]):
        def __call__(
            self,
            callable: (
                Callable[P, R]
                | Callable[Concatenate[type[T], P], R]
                # | Callable[Concatenate[Any, P], R] # Don't use, breaks stuff.
                | Callable[[type[T]], R]
                | Descriptor
            ),
        ) -> FunctionWrapper[P, R]: ...

    class PartialFunctionDecorator:
        @overload
        def __call__(
            self, wrapper: GenericCallableWrapperFunction[P, R], /
        ) -> FunctionDecorator[P, R]: ...
        @overload
        def __call__(
            self, wrapper: ClassMethodWrapperFunction[P, R], /
        ) -> FunctionDecorator[P, R]: ...
        @overload
        def __call__(
            self, wrapper: InstanceMethodWrapperFunction[P, R], /
        ) -> FunctionDecorator[P, R]: ...

    # ... Decorator applied to class type.

    @overload
    def decorator(wrapper: type[T], /) -> FunctionDecorator[Any, Any]: ...

    # ... Decorator applied to function or method.

    @overload
    def decorator(
        wrapper: GenericCallableWrapperFunction[P, R], /
    ) -> FunctionDecorator[P, R]: ...
    @overload
    def decorator(
        wrapper: ClassMethodWrapperFunction[P, R], /
    ) -> FunctionDecorator[P, R]: ...
    @overload
    def decorator(
        wrapper: InstanceMethodWrapperFunction[P, R], /
    ) -> FunctionDecorator[P, R]: ...

    # ... Positional arguments.

    @overload
    def decorator(
        *,
        enabled: bool | Boolean | Callable[[], bool] | None = None,
        adapter: str | FullArgSpec | AdapterFactory | Callable[..., Any] | None = None,
        proxy: type[FunctionWrapper[Any, Any]] | None = None,
    ) -> PartialFunctionDecorator: ...

    # function_wrapper()

    @overload
    def function_wrapper(wrapper: type[Any]) -> FunctionDecorator[Any, Any]: ...
    @overload
    def function_wrapper(
        wrapper: GenericCallableWrapperFunction[P, R],
    ) -> FunctionDecorator[P, R]: ...
    @overload
    def function_wrapper(
        wrapper: ClassMethodWrapperFunction[P, R],
    ) -> FunctionDecorator[P, R]: ...
    @overload
    def function_wrapper(
        wrapper: InstanceMethodWrapperFunction[P, R],
    ) -> FunctionDecorator[P, R]: ...
    # @overload
    # def function_wrapper(wrapper: Any) -> FunctionDecorator[Any, Any]: ... # Don't use, breaks stuff.

    # wrap_function_wrapper()

    def wrap_function_wrapper(
        target: ModuleType | type[Any] | Any | str,
        name: str,
        wrapper: WrapperFunction[P, R],
    ) -> FunctionWrapper[P, R]: ...

    # patch_function_wrapper()

    class WrapperDecorator:
        def __call__(self, wrapper: WrapperFunction[P, R]) -> FunctionWrapper[P, R]: ...

    def patch_function_wrapper(
        target: ModuleType | type[Any] | Any | str,
        name: str,
        enabled: bool | Boolean | Callable[[], bool] | None = None,
    ) -> WrapperDecorator: ...

    # transient_function_wrapper()

    class TransientDecorator:
        def __call__(
            self, wrapper: WrapperFunction[P, R]
        ) -> FunctionDecorator[P, R]: ...

    def transient_function_wrapper(
        target: ModuleType | type[Any] | Any | str, name: str
    ) -> TransientDecorator: ...

    # resolve_path()

    def resolve_path(
        target: ModuleType | type[Any] | Any | str, name: str
    ) -> tuple[ModuleType | type[Any] | Any, str, Callable[..., Any]]: ...

    # apply_patch()

    def apply_patch(
        parent: ModuleType | type[Any] | Any,
        attribute: str,
        replacement: Any,
    ) -> None: ...

    # wrap_object()

    WrapperFactory = Callable[
        [Callable[..., Any], tuple[Any, ...], dict[str, Any]], type[ObjectProxy[Any]]
    ]

    def wrap_object(
        target: ModuleType | type[Any] | Any | str,
        name: str,
        factory: WrapperFactory | type[ObjectProxy[Any]],
        args: tuple[Any, ...],
        kwargs: dict[str, Any],
    ) -> Any: ...

    # wrap_object_attribute()

    def wrap_object_attribute(
        target: ModuleType | type[Any] | Any | str,
        name: str,
        factory: WrapperFactory | type[ObjectProxy[Any]],
        args: tuple[Any, ...] = (),
        kwargs: dict[str, Any] = {},
    ) -> Any: ...

    # register_post_import_hook()

    def register_post_import_hook(
        hook: Callable[[ModuleType], Any] | str, name: str
    ) -> None: ...

    # discover_post_import_hooks()

    def discover_post_import_hooks(group: str) -> None: ...

    # notify_module_loaded()

    def notify_module_loaded(module: ModuleType) -> None: ...

    # when_imported()

    class ImportHookDecorator:
        def __call__(self, hook: Callable[[ModuleType], Any]) -> Callable[..., Any]: ...

    def when_imported(name: str) -> ImportHookDecorator: ...

    # synchronized()

    class SynchronizedObject:
        def __call__(self, wrapped: Callable[P, R]) -> Callable[P, R]: ...
        def __enter__(self) -> Any: ...
        def __exit__(
            self,
            exc_type: type[BaseException] | None,
            exc_value: BaseException | None,
            traceback: TracebackType | None,
        ) -> bool | None: ...

    @overload
    def synchronized(wrapped: Callable[P, R]) -> Callable[P, R]: ...
    @overload
    def synchronized(wrapped: Any) -> SynchronizedObject: ...
