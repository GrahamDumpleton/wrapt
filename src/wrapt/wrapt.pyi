from typing import Any, Generic, TypeVar, Callable, Optional

F = TypeVar('F', bound=Callable[..., Any])
A = TypeVar('A', bound=Callable[..., Any])
T = TypeVar("T", bound=Any)

def decorator(wrapper: F, enabled: Optional[bool] = None, adapter: Optional[A] = None) -> F: ...

class ObjectProxy(Generic[T]):
    __wrapped__ : T

    def __init__(self, wrapped: T):
        ...
