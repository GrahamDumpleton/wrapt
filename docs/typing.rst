Type Hinting
============

As of version 2.0.0, **wrapt** provides inline type hints for its public APIs to
improve interoperability with static type checkers (e.g. ``mypy``). The type
metadata is available when running on Python 3.10 or later (the minimum version
the annotations target).

The annotations aim to be as precise as practical in order to maximise
inference quality and surface errors early. In some situations you may still
need to add explicit annotations yourself, sometimes by introducing small
helper wrapper functions to guide the type checker.

Certain advanced decorator and wrapper patterns enabled by **wrapt** are simply
too dynamic to express exactly in today's static type systems, so exhaustive
type checking cannot be applied there. The principal categories of limitations,
their underlying causes, and recommended workarounds are described in the
sections which follow.

Function Decorators
-------------------

The **wrapt** module exposes two helpers for authoring function decorators:
``@decorator`` and ``@function_wrapper``. ``@function_wrapper`` is a lightweight
subset of ``@decorator`` intended for straightforward function wrapping; it
omits the advanced extension points in exchange for a smaller, simpler
surface. We will use ``@decorator`` in this discussion of type hints, but
``@function_wrapper`` is also fully type hinted and can be used in a similar way.

The most basic example of a function decorator is:

::

    @wrapt.decorator
    def pass_through(wrapped, instance, args, kwargs):
        return wrapped(*args, **kwargs)

    @pass_through
    def add(a, b):
        return a + b

    result = add(2, 3)

In this example, because no type hints are provided, the type checker will not
be able to infer the types of the function's parameters or return value.

Adding type hints, the example becomes:

::

    from typing import Any, Callable

    @wrapt.decorator  # <-- Type checking applied here.
    def pass_through(
        wrapped: Callable[[int, int], int],
        instance: Any,
        args: tuple[Any, ...],
        kwargs: dict[str, Any],
    ) -> int:
        return wrapped(*args, **kwargs)

    @pass_through  # <-- Type checking not applied here (see below).
    def add(a: int, b: int) -> int:
        return a + b

    result: int = add(2, 3)  # <-- Type checking applied here.

Annotate ``wrapped`` with a ``Callable`` whose signature matches the functions you
intend to decorate. Leave ``instance`` as ``Any`` in most cases; if you need to be
stricter you can use ``Any | type[Any] | None``. You can narrow ``args`` / ``kwargs``
based on the accepted parameter types, but doing so rarely improves type
checking and usually adds noise, so the generic ``tuple[Any, ...]`` and
``dict[str, Any]`` forms are normally sufficient.

The wrapper's return type annotation should mirror the return type of the
wrapped function.

With the type hints added, the type checker can now validate both the
arguments supplied to the decorated function and its return value at the point
it is being called. For example, the type checker will flag incorrect argument
types or an incompatible assigned result as in the following.

::

    result: str = add("hello", "world")  # Error: incompatible types

Because a single **wrapt** decorator can be applied to both plain functions
and bound or class methods, there are practical limits to how precisely we can
match the call signatures. In particular, the implicit ``self`` / ``cls`` binding
for methods cannot be reconciled reliably with the standalone function
signature you must supply for ``wrapped``. While partial workarounds are
possible, they introduce fragility and false positives, so for now type
checking has been disabled at the point where the decorator is being applied
to a method or function.

Argument Matching
-----------------

Functions being decorated can have default arguments, and so long as the
decorator is aware of them, it can handle them correctly at runtime.
Unfortunately due to limitations in the type system, a type checker doesn't
recognise the default arguments of the wrapped function and so will indicate
an error when you don't still explicitly provide the argument, even though the
call will work fine at runtime.

::

    from typing import Any, Callable

    @wrapt.decorator
    def pass_through(
        wrapped: Callable[[int, int], int],
        instance: Any,
        args: tuple[Any, ...],
        kwargs: dict[str, Any],
    ) -> int:
        return wrapped(*args, **kwargs)

    @pass_through
    def add(a: int, b: int = 0) -> int:
        return a + b

    result1: int = add(2, 3)

    # Error: type checker doesn't recognise default arguments.

    result2: int = add(2) # <-- Invalid error warning.

Type checkers also have issue with using keyword parameters, thus they will
wrongly flag use of keyword arguments even though at runtime it still works.

::

     # Error: type checker doesn't recognise keyword arguments.

    result3: int = add(2, b=3) # <-- Invalid error warning.

If using these patterns and you don't like see the errors, you will need to
flag the type checker to ignore them. For example, with ``mypy`` you can use
``# type: ignore`` comments to suppress the warnings:

Signature Adapters
------------------

Sometimes you want the decorated callable to present a different public
signature from the underlying implementation (for example, to narrow the
parameters, rename them, or enforce keyword-only usage). You can express this
using a signature adapter: a small prototype function whose only purpose is
to declare the outward-facing signature the wrapped function should appear to
have after decoration.

For instance, imagine the original function returns an integer, but you want
the decorated function to return a string. You would define a signature adapter
like this:

::

    def adapter_prototype(i: int) -> str: ...

    @wrapt.decorator(adapter=wrapt.adapter_factory(adapter_prototype))
    def int_to_str(wrapped, instance, args, kwargs):
        return str(wrapped(*args, **kwargs))

    @int_to_str
    def function(x) -> int:
        """A function that takes an integer and returns it."""
        return x

    result = function(1)

In this example we passed the prototype function itself via the ``adapter``
argument. **wrapt** also supports alternative forms: you can supply the
prototype as a string, or return a pre-formatted argument spec instead of a
callable.

Declaring the adapter explicitly ensures that runtime introspection
(``inspect.signature``, ``help()``, IDE tooling, etc.) reports the adapted
signature rather than the underlying implementation detail. Because the
adaptation is applied dynamically (and the prototype may itself be generated
at runtime), **wrapt** cannot reliably infer the target signature from the
wrapped function alone, and so you must provide it if you want accurate type
checking.

::

    def adapter_prototype(i: int) -> str: ...

    def int_to_str(wrapped: Callable[[int], int]) -> Callable[[int], str]:
        @wrapt.decorator(adapter=wrapt.adapter_factory(adapter_prototype))
        def wrapper(
            wrapped: Callable[..., Any],
            instance: Any,
            args: tuple[Any, ...],
            kwargs: dict[str, Any],
        ) -> Any:
            return str(wrapped(*args, **kwargs))

        return wrapper(wrapped)

    @int_to_str
    def function(x: int) -> int:
        """A function that takes an integer and returns it."""
        return x

    result: str = function(1)

In this version we introduced an outer helper that constructs the decorator and
added explicit type hints to its parameters and return type. This allows the
type checker to validate calls to the decorated function and propagate the
correct return type.

Note that inside the decorator body the ``wrapped`` callable is annotated
as accepting any arguments and returning ``Any``. You could just as well omit
those inner annotations as what matters for most static checking is the
user facing signature exposed by the outer wrapper.

Decorating Classes
------------------

Decorators can be applied to classes as well as functions and methods. when
applied to a class, the decorator object effectively replaces the original.
With the way the **wrapt** decorator works, it is still possible to use the
decorated class as a base class in an inheritance hierarchy, however, this
confuses the type checker.

::

    @pass_through
    class BaseClass:
        def __init__(self): ...

    # Error: type checker doesn't recognise the class as a base class.

    class DerivedClass(BaseClass):  # <-- Invalid error warning.
        def __init__(self): ...

The type checker can also give invalid error warnings when using functions
such as `issubclass()` due to not recognising the decorated class as a
class type.

::

    # Error: type checker doesn't recognise the class as a base class.

    issubclass(DerivedClass, BaseClass) # <-- Invalid error warning.

Class as Decorator
------------------

Normally decorators are functions, but it is also possible to use a class as a
decorator. In this situation the wrapper function (``__call__()`` method of class)
is not type checked as it would be if the ``@decorator`` were being applied to it
directly. Further, the type checker cannot match the arguments for the
constructor of the class at the point it it is being created.

::

    @wrapt.decorator
    class ClassDecorator:
        def __init__(self, arg: str): ...

        # Error: type checker will not check arguments of wrapper function.

        def __call__(self, wrapped, instance, args, kwargs): ... # <-- Not checked.

    # Error: type checker doesn't recognise arguments correctly.

    @ClassDecorator("string") # <-- Invalid error warning.
    def function(): ...

Object Proxy
------------

Due to the magic of how the ``ObjectProxy`` class in **wrapt** works, you may find
that the type checker will generate errors about using it and the above are not
all the possible issues you may encounter. To allow for further investigation
and improvement of the type hints, please report any issues you find with
using **wrapt**. For more notable cases we can add at least add additional
documentation here with warnings or workarounds.
