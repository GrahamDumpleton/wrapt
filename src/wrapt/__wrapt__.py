"""This module is used to switch between C and Python implementations of the
wrappers.
"""

import os

from .wrappers import (
    BoundFunctionWrapper,
    CallableObjectProxy,
    FunctionWrapper,
    ObjectProxy,
    PartialCallableObjectProxy,
    _FunctionWrapperBase,
)

# Import Python fallbacks.


# Try to use C extensions if not disabled.

_use_extensions = not os.environ.get("WRAPT_DISABLE_EXTENSIONS")

if _use_extensions:
    try:
        from ._wrappers import (  # type: ignore[no-redef,import-not-found]
            BoundFunctionWrapper,
            CallableObjectProxy,
            FunctionWrapper,
            ObjectProxy,
            PartialCallableObjectProxy,
            _FunctionWrapperBase,
        )
    except ImportError:
        # C extensions not available, using Python implementations
        pass

# Provide an alias for partial().


def partial(*args, **kwargs):
    """Create a callable object proxy with partial application of the given
    arguments and keywords. This behaves the same as `functools.partial`, but
    implemented using the `ObjectProxy` class to provide better support for
    introspection.
    """
    return PartialCallableObjectProxy(*args, **kwargs)
