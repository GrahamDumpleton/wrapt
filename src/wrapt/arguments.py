"""The inspect.formatargspec() function was dropped in Python 3.11 but we need
it for when constructing signature changing decorators based on result of
inspect.getfullargspec(). The code here implements inspect.formatargspec() based
on Parameter and Signature from inspect module, which were added in Python 3.6.
Thanks to Cyril Jouve for the implementation.
"""

from typing import Any, Callable, List, Mapping, Optional, Sequence, Tuple

try:
    from inspect import Parameter, Signature
except ImportError:
    from inspect import formatargspec  # type: ignore[attr-defined]
else:

    def formatargspec(
        args: List[str],
        varargs: Optional[str] = None,
        varkw: Optional[str] = None,
        defaults: Optional[Tuple[Any, ...]] = None,
        kwonlyargs: Optional[Sequence[str]] = None,
        kwonlydefaults: Optional[Mapping[str, Any]] = None,
        annotations: Mapping[str, Any] = {},
        formatarg: Callable[[str], str] = str,
        formatvarargs: Callable[[str], str] = lambda name: "*" + name,
        formatvarkw: Callable[[str], str] = lambda name: "**" + name,
        formatvalue: Callable[[Any], str] = lambda value: "=" + repr(value),
        formatreturns: Callable[[Any], str] = lambda text: " -> " + text,
        formatannotation: Callable[[Any], str] = lambda annot: " -> " + repr(annot),
    ) -> str:
        if kwonlyargs is None:
            kwonlyargs = ()
        if kwonlydefaults is None:
            kwonlydefaults = {}
        ndefaults = len(defaults) if defaults else 0
        parameters = [
            Parameter(
                arg,
                Parameter.POSITIONAL_OR_KEYWORD,
                default=defaults[i] if defaults and i >= 0 else Parameter.empty,
                annotation=annotations.get(arg, Parameter.empty),
            )
            for i, arg in enumerate(args, ndefaults - len(args))
        ]
        if varargs:
            parameters.append(Parameter(varargs, Parameter.VAR_POSITIONAL))
        parameters.extend(
            Parameter(
                kwonlyarg,
                Parameter.KEYWORD_ONLY,
                default=kwonlydefaults.get(kwonlyarg, Parameter.empty),
                annotation=annotations.get(kwonlyarg, Parameter.empty),
            )
            for kwonlyarg in kwonlyargs
        )
        if varkw:
            parameters.append(Parameter(varkw, Parameter.VAR_KEYWORD))
        return_annotation = annotations.get("return", Signature.empty)
        return str(Signature(parameters, return_annotation=return_annotation))
