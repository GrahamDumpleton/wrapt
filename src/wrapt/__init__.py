"""
Wrapt is a library for decorators, wrappers and monkey patching.
"""

__version_info__ = ("2", "0", "0", "rc2")
__version__ = ".".join(__version_info__)

from .__wrapt__ import (
    BoundFunctionWrapper,
    CallableObjectProxy,
    FunctionWrapper,
    ObjectProxy,
    PartialCallableObjectProxy,
    partial,
)
from .decorators import AdapterFactory, adapter_factory, decorator, synchronized
from .importer import (
    discover_post_import_hooks,
    notify_module_loaded,
    register_post_import_hook,
    when_imported,
)
from .patches import (
    apply_patch,
    function_wrapper,
    patch_function_wrapper,
    resolve_path,
    transient_function_wrapper,
    wrap_function_wrapper,
    wrap_object,
    wrap_object_attribute,
)
from .weakrefs import WeakFunctionProxy

__all__ = (
    "BoundFunctionWrapper",
    "CallableObjectProxy",
    "FunctionWrapper",
    "ObjectProxy",
    "PartialCallableObjectProxy",
    "partial",
    "AdapterFactory",
    "adapter_factory",
    "decorator",
    "synchronized",
    "discover_post_import_hooks",
    "notify_module_loaded",
    "register_post_import_hook",
    "when_imported",
    "apply_patch",
    "function_wrapper",
    "patch_function_wrapper",
    "resolve_path",
    "transient_function_wrapper",
    "wrap_function_wrapper",
    "wrap_object",
    "wrap_object_attribute",
    "WeakFunctionProxy",
)
