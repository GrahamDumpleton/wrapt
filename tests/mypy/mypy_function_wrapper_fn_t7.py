"""
This example demonstrates usage of the function_wrapper() function.
"""

from typing import Any, Callable, Concatenate, ParamSpec, TypeVar

from wrapt import function_wrapper

P = ParamSpec("P")
R = TypeVar("R")


@function_wrapper
def wrapper1(
    wrapped: Callable[P, R],
    instance: Any,
    args: tuple[Any, ...],
    kwargs: dict[str, Any],
) -> R:
    return wrapped(*args, **kwargs)


@wrapper1
def function1(x: int, y: str = "string") -> str:
    """A simple function to be wrapped."""
    return f"{x}: {y}"


# No arguments. (FAIL)
function1()

function1(1, "test")
function1(2)
function1(3, y="override")


class ExampleClass1:
    """A class with methods to be wrapped."""

    def __init__(
        self,
        value: int,
    ) -> None:
        self.value = value

    @wrapper1
    def __call__(self, value: int, name: str) -> str:
        return f"callable: {value}"

    @wrapper1
    def instance_method(self, value: int, name: str) -> str:
        return f"instance: {value}"

    @wrapper1
    @classmethod
    def class_method(cls, value: int, name: str) -> str:
        return f"class: {value}"

    @wrapper1
    @staticmethod
    def static_method(value: int, name: str) -> str:
        return f"static: {value}"


example_instance1 = ExampleClass1(0)

# No arguments. (FAIL)
example_instance1.instance_method()

# No arguments. (FAIL)
example_instance1.class_method()

# No arguments. (FAIL)
example_instance1.static_method()

example_instance1.instance_method(1, "test")
example_instance1.class_method(1, "test")
example_instance1.static_method(1, "test")

example_instance1(1, "test")

# No arguments. (FAIL)
ExampleClass1.class_method()

# No arguments. (FAIL)
ExampleClass1.static_method()

ExampleClass1.class_method(1, "test")
ExampleClass1.static_method(1, "test")


@function_wrapper
def wrapper2(
    wrapped: Callable[..., Any],
    instance: Any,
    args: tuple[Any, ...],
    kwargs: dict[str, Any],
) -> Any:
    return wrapped(*args, **kwargs)


@wrapper2
def function2(x: int, y: str | None = None) -> str:
    """A simple function to be wrapped."""
    if y is None:
        y = "default"
    return f"{x}: {y}"


function2()


function2(1, "test")
function2(2)
function2(3, y="override")


class ExampleClass2:
    """A class with methods to be wrapped."""

    def __init__(self, value: int) -> None:
        self.value = value

    @wrapper2
    def __call__(self, value: int) -> str:
        return f"callable: {value}"

    @wrapper2
    def instance_method(self, value: int) -> str:
        return f"instance: {value}"

    @wrapper2
    @classmethod
    def class_method(cls, value: int) -> str:
        return f"class: {value}"

    @wrapper2
    @staticmethod
    def static_method(value: int) -> str:
        return f"static: {value}"


example_instance2 = ExampleClass2(0)

example_instance2.instance_method(1)
example_instance2.class_method(1)
example_instance2.static_method(1)

example_instance2(1)


example_instance2.instance_method()
example_instance2.class_method()
example_instance2.static_method()

example_instance2()


class DecoratorClass:
    @function_wrapper
    def wrapper(
        self,
        wrapped: Callable[P, R],
        instance: Any,
        args: tuple[Any, ...],
        kwargs: dict[str, Any],
    ) -> R:
        return wrapped(*args, **kwargs)


decorator_class = DecoratorClass()


@decorator_class.wrapper
def function3(x: int, y: str | None = None) -> str:
    """A simple function to be wrapped."""
    if y is None:
        y = "default"
    return f"{x}: {y}"


# No arguments. (FAIL)
function3()

function3(1, "test")
function3(2)
function3(3, y="override")
