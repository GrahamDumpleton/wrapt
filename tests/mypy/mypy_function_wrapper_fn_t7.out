tests/mypy/mypy_function_wrapper_fn_t7.py:32: error: Argument 1 to "__call__" of "FunctionWrapper" has incompatible type "int"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:32: error: Argument 2 to "__call__" of "FunctionWrapper" has incompatible type "str"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:33: error: Argument 1 to "__call__" of "FunctionWrapper" has incompatible type "int"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:34: error: Argument 1 to "__call__" of "FunctionWrapper" has incompatible type "int"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:34: error: Argument "y" to "__call__" of "FunctionWrapper" has incompatible type "str"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:76: error: Argument 1 to "__call__" of "BoundFunctionWrapper" has incompatible type "int"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:76: error: Argument 2 to "__call__" of "BoundFunctionWrapper" has incompatible type "str"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:77: error: Argument 1 to "__call__" of "BoundFunctionWrapper" has incompatible type "int"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:77: error: Argument 2 to "__call__" of "BoundFunctionWrapper" has incompatible type "str"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:78: error: Argument 1 to "__call__" of "BoundFunctionWrapper" has incompatible type "int"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:78: error: Argument 2 to "__call__" of "BoundFunctionWrapper" has incompatible type "str"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:80: error: Argument 1 to "__call__" of "BoundFunctionWrapper" has incompatible type "int"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:80: error: Argument 2 to "__call__" of "BoundFunctionWrapper" has incompatible type "str"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:88: error: Argument 1 to "__call__" of "BoundFunctionWrapper" has incompatible type "int"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:88: error: Argument 2 to "__call__" of "BoundFunctionWrapper" has incompatible type "str"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:89: error: Argument 1 to "__call__" of "BoundFunctionWrapper" has incompatible type "int"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:89: error: Argument 2 to "__call__" of "BoundFunctionWrapper" has incompatible type "str"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:185: error: Argument 1 to "__call__" of "FunctionWrapper" has incompatible type "int"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:185: error: Argument 2 to "__call__" of "FunctionWrapper" has incompatible type "str"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:186: error: Argument 1 to "__call__" of "FunctionWrapper" has incompatible type "int"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:187: error: Argument 1 to "__call__" of "FunctionWrapper" has incompatible type "int"; expected "Never"  [arg-type]
tests/mypy/mypy_function_wrapper_fn_t7.py:187: error: Argument "y" to "__call__" of "FunctionWrapper" has incompatible type "str"; expected "Never"  [arg-type]
Found 22 errors in 1 file (checked 1 source file)
